tempesta_fw/tls.c

```diff
@@ -379,23 +388,87 @@ tfw_tls_encrypt(struct sock *sk, struct sk_buff *skb, unsigned int limit)
 
 	if (likely(sgt.nents <= AUTO_SEGS_N)) {
 		sgt.sgl = sg;
+		out_sgt.sgl = out_sg;
+		pages = auto_pages;
 	} else {
 		sgt.sgl = kmalloc(sizeof(struct scatterlist) * sgt.nents,
 				  GFP_ATOMIC);
 		if (!sgt.sgl) {
 			T_WARN("cannot alloc TLS encryption scatter list.\n");
 			return -ENOMEM;
 		}
+		out_sgt.sgl = kmalloc(sizeof(struct scatterlist) * out_sgt.nents,
+				      GFP_ATOMIC);
+		if (!out_sgt.sgl) {
+			T_WARN("cannot alloc TLS encryption scatter list.\n");
+			return -ENOMEM;
+		}
+		pages = kmalloc(sizeof(struct page *) * out_sgt.nents,
+				GFP_ATOMIC);
+		if (!pages) {
+			T_WARN("cannot alloc memory.\n");
+			return -ENOMEM;
+		}
 	}
 	sg_init_table(sgt.sgl, sgt.nents);
+	sg_init_table(out_sgt.sgl, out_sgt.nents);
 
-	for (next = skb, frags = 0; ; ) {
+	remain = offset = 0;
+	for (next = skb, frags = 0, out_frags = 0, elt = 0; ; ) {
 		/*
 		 * skb data and tails are already adjusted above,
 		 * so use zero offset and skb->len.
 		 */
 		r = skb_to_sgvec(next, sgt.sgl + frags, 0, next->len);
 
+		if (skb_shinfo(next)->tx_flags & SKBTX_SHARED_FRAG ||
+		    tls->resp_from_cache)
+		{
+			sg_set_buf(out_sgt.sgl + out_frags, next->data,
+			           skb_headlen(next));
+			out_frags++;
+
+			for (i = 0; i < skb_shinfo(next)->nr_frags; i++) {
+				skb_frag_t *f = &skb_shinfo(next)->frags[i];
+				unsigned int size;
+				struct page *p;
+
+				size = skb_frag_size(f);
+				if (remain < size) {
+					int order = get_order(size);
+					p = alloc_pages(GFP_ATOMIC, order);
+					if (!p) {
+						r = -ENOMEM;
+						goto out;
+					}
+					remain = (1 << order) * PAGE_SIZE;
+					offset = 0;
+				} else {
+					skb_frag_t *prev_f;
+
+					prev_f = &skb_shinfo(next)->frags[i - 1];
```

Comment:
Say, we processed one `skb`, and after the last fragment of that `skb` there is remaining room in an allocated page. Then we pick another skb (`next = tcp_write_queue_next(sk, next);`), and start the cycle again. This time `remain` may be large enough so we get to this line with `i` being `0`. With subsequent out-of-bounds access.